File: lib/zexray/trace_log.ex
defmodule Zexray.TraceLog do
  @moduledoc """
  Trace Log
  """

  use Zexray.Enum

  alias Zexray.NIF

  @doc """
  Show trace log messages (LOG_DEBUG, LOG_INFO, LOG_WARNING, LOG_ERROR...)
  """
  @spec trace_log(
          log_level :: Zexray.Enum.TraceLogLevel.t(),
          text :: binary
        ) :: :ok
  if Application.compile_env(:zexray, :trace_log, true) do
    defdelegate trace_log(
                  log_level,
                  text
                ),
                to: NIF,
                as: :trace_log
  else
    def trace_log(_log_level, _text), do: :ok
  end

  @doc """
  Trace logging, intended for internal use only

  Show `:trace` log message
  """
  @spec trace(text :: binary) :: :ok
  if Application.compile_env(:zexray, :trace_log, true) do
    def trace(text) do
      trace_log(enum_trace_log_level(:trace), text)
    end
  else
    def trace(_text), do: :ok
  end

  @doc """
  Debug logging, used for internal debugging, it should be disabled on release builds

  Show `:debug` log message
  """
  @spec debug(text :: binary) :: :ok
  if Application.compile_env(:zexray, :trace_log, true) and
       Application.compile_env(:zexray, :trace_log_debug, false) do
    def debug(text) do
      trace_log(enum_trace_log_level(:debug), text)
    end
  else
    def debug(_text), do: :ok
  end

  @doc """
  Info logging, used for program execution info

  Show `:info` log message
  """
  @spec info(text :: binary) :: :ok
  if Application.compile_env(:zexray, :trace_log, true) do
    def info(text) do
      trace_log(enum_trace_log_level(:info), text)
    end
  else
    def info(_text), do: :ok
  end

  @doc """
  Warning logging, used on recoverable failures 

  Show `:warning` log message
  """
  @spec warning(text :: binary) :: :ok
  if Application.compile_env(:zexray, :trace_log, true) do
    def warning(text) do
      trace_log(enum_trace_log_level(:warning), text)
    end
  else
    def warning(_text), do: :ok
  end

  @doc """
  Error logging, used on unrecoverable failures 

  Show `:error` log message
  """
  @spec error(text :: binary) :: :ok
  if Application.compile_env(:zexray, :trace_log, true) do
    def error(text) do
      trace_log(enum_trace_log_level(:error), text)
    end
  else
    def error(_text), do: :ok
  end

  @doc """
  Fatal logging, used to abort program: exit(EXIT_FAILURE) 

  Show `:fatal` log message
  """
  @spec fatal(text :: binary) :: :ok
  if Application.compile_env(:zexray, :trace_log, true) do
    def fatal(text) do
      trace_log(enum_trace_log_level(:fatal), text)
    end
  else
    def fatal(_text), do: :ok
  end

  @doc """
  Set the current threshold (minimum) log level
  """
  @spec set_level(log_level :: Zexray.Enum.TraceLogLevel.t()) :: :ok
  defdelegate set_level(log_level), to: NIF, as: :set_trace_log_level
end

File: src/nifs/trace_log.zig
const std = @import("std");
const assert = std.debug.assert;
const e = @import("../erl_nif.zig");
const rl = @import("../raylib.zig");

const core = @import("../core.zig");

const stdlib = @cImport({
    @cInclude("stdlib.h");
});

const build_config = @import("config");

var log_type_level = rl.LOG_INFO;

pub const exported_nifs = [_]e.ErlNifFunc{
    // TraceLog
    .{ .name = "trace_log", .arity = 2, .fptr = core.nif_wrapper(nif_trace_log), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "set_trace_log_level", .arity = 1, .fptr = core.nif_wrapper(nif_set_trace_log_level), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "set_trace_log_callback", .arity = 0, .fptr = core.nif_wrapper(nif_set_trace_log_callback), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
};

////////////////
//  TraceLog  //
////////////////

pub fn trace_log(log_level: c_int, text: []const u8) void {
    if (build_config.trace_log) {
        if (log_level >= log_type_level) {
            const writer = std.io.getStdErr().writer();

            switch (log_level) {
                rl.LOG_TRACE => writer.writeAll("TRACE:   ") catch return,
                rl.LOG_DEBUG => writer.writeAll("DEBUG:   ") catch return,
                rl.LOG_INFO => writer.writeAll("INFO:    ") catch return,
                rl.LOG_WARNING => writer.writeAll("WARNING: ") catch return,
                rl.LOG_ERROR => writer.writeAll("ERROR:   ") catch return,
                rl.LOG_FATAL => writer.writeAll("FATAL:   ") catch return,
                else => return,
            }

            writer.writeAll(text) catch return;
            writer.writeAll("\n\r") catch return;

            if (log_level == rl.LOG_FATAL) {
                stdlib.exit(stdlib.EXIT_FAILURE);
            }
        }
    }
}

pub fn traceLog(log_level: c_int, text: [*c]const u8, args: *anyopaque) callconv(.C) void {
    const buf_len = rl.MAX_TRACELOG_MSG_LENGTH * 4;
    var buf: [buf_len]u8 = std.mem.zeroes([buf_len]u8);

    const len: usize = @intCast(rl.vsnprintf(&buf, buf_len, text, @ptrCast(@alignCast(args))));

    trace_log(log_level, buf[0..len]);
}

/// Show trace log messages (LOG_DEBUG, LOG_INFO, LOG_WARNING, LOG_ERROR...)
///
/// raylib.h
/// RLAPI void TraceLog(int logLevel, const char *text, ...);
fn nif_trace_log(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 2);

    // Arguments

    const log_level: c_int = core.Int.get(env, argv[0]) catch {
        return error.invalid_argument_log_level;
    };

    const arg_text = core.ArgumentBinary(core.CString, rl.allocator).get(env, argv[1]) catch {
        return error.invalid_argument_text;
    };
    defer arg_text.free();
    const text = arg_text.data;

    // Function

    trace_log(log_level, text);

    // Return

    return core.Atom.make(env, "ok");
}

/// Set the current threshold (minimum) log level
///
/// raylib.h
/// RLAPI void SetTraceLogLevel(int logLevel);
fn nif_set_trace_log_level(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const log_level: c_int = core.Int.get(env, argv[0]) catch {
        return error.invalid_argument_log_level;
    };

    // Function

    rl.SetTraceLogLevel(log_level);
    log_type_level = log_level;

    // Return

    return core.Atom.make(env, "ok");
}

/// Set custom trace log
///
/// raylib.h
/// RLAPI void SetTraceLogCallback(TraceLogCallback callback);
fn nif_set_trace_log_callback(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.SetTraceLogCallback(@ptrCast(&traceLog));

    // Return

    return core.Atom.make(env, "ok");
}
